opm/simulators/linalg/cuistl/CuSeqILU0.cpp

```diff
@@ -0,0 +1,362 @@
+/*
+  Copyright SINTEF AS 2022
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <cusparse.h>
+#include <dune/common/fmatrix.hh>
+#include <dune/common/fvector.hh>
+#include <dune/istl/bcrsmatrix.hh>
+#include <dune/istl/bvector.hh>
+#include <opm/common/ErrorMacros.hpp>
+#include <opm/simulators/linalg/cuistl/CuSeqILU0.hpp>
+#include <opm/simulators/linalg/cuistl/impl/cusparse_constants.hpp>
+#include <opm/simulators/linalg/cuistl/impl/cusparse_safe_call.hpp>
+#include <opm/simulators/linalg/cuistl/impl/cusparse_wrapper.hpp>
+#include <opm/simulators/linalg/cuistl/impl/fix_zero_diagonal.hpp>
+#include <opm/simulators/linalg/cuistl/time_to_file.hpp>
+#include <opm/simulators/linalg/matrixblock.hh>
+
+
+
+namespace Opm::cuistl
+{
+
+template <class M, class X, class Y, int l>
+CuSeqILU0<M, X, Y, l>::CuSeqILU0(const M& A, scalar_field_type w)
+    : m_underlyingMatrix(A)
+    , m_w(w)
+    , m_LU(CuSparseMatrix<field_type>::fromMatrix(impl::makeMatrixWithNonzeroDiagonal(A)))
```

Comment:
Will this change the original matrix? (I recall a function doing that with a const_cast. Not sure whether it was this) In that case do we really want this?